src_blog/_posts/Grain-Formatter.md

```diff
@@ -0,0 +1,120 @@
+
+```sh
+(a,b) and [1,2,3] and  {name:"Super",    value:"Grain"}
+```
+
+becomes
+```sh
+(a, b) and [1, 2, 3] and { name: "Super", value: "Grain" } 
+```
+
+or if it split over lines we add a trailing the comma:
+
+```sh
+enum Animals { 
+  Dog, 
+  Cat,
+}
+```
+
+
+We apply some smaller styling choices too, for example a single argument function isn’t wrapped with parenthesis. 
+
+```sh
+let add1 = x => x + 1
+let add = (x, y) => x + y
+```
+
+Also quite important is the ability to tag code as not to be formatted
+
+```sh
+// formatter-ignore
+```
+
+Use the formatter-ignore annotation in a comment and the following statement will be reproduced as-is.
+
+# Implementation
+
+The formatter works by using the Grain compiler to first parse the source code into what's called an Abstract Syntax Tree or AST that captures the meaning of the program as a tree of patterns and expressions.    Then we just write the tree back out as valid Grain code.  That's sometimes easier said than done, and here are some of the challenges we faced.
+
+
+## Challenges
+
+-	Comments.   Comments are by definition ignored when the compiler is comprehending the meaning of our code and so isn't stored in the AST.  We need to take the AST and the comments extracted from the source, bring the regenerated code and comments back together and check how the code looks with comments when formatting.   In some of the more esoteric applications of comments, their placement can be ambiguous in relation to source code that also doesn’t appear in the AST, for example
+
+```sh
+variable1 /* comment */ , variable2
+```
+
+We know the locations (source code ranges) of the variable names and the comment, but we don’t know if the comment comes before or after the comma.
+
+-	We use various forms of syntactic sugar.  One of my favourites is converting list syntax of [1,2,3] into repeated application of the list cons operator.    The formatter needs to be aware of these and rewrite the AST back out in sugared syntax.
+
+- Blanks lines are useful for structuring and organising code but also need to be styled for consistency.  Removing all blank lines really impacts on how the source code looks and reads.    We chose to compress multiple blank lines down into a single blank line, but to always retain at least one blank line wherever one was found.
+
+-	Arguments over style.  As you might imagine we had quite a lot of discussion about style choices.  There really is no true way that suits everyone, and whatever is chosen will always be a compromise.   However what we wanted to avoid was adding configurable formatting where different formatting choices can be specified.  This really undoes the premise of all formatted Grain code looking the same, which we think really helps all Grain programmers, new or experienced, to look at Grain code and find it easy to read through because of the consistency, be than in end user code or the standard libraries.
```

Comment:
```suggestion
-	Arguments over style. As you might imagine we had quite a lot of discussion about style choices. There really is no single way that suits everyone, and whatever is chosen will always be a compromise. However, what we wanted to avoid was adding configurable formatting where different formatting choices can be specified.  This really undermines the premise of all formatted Grain code looking the same, which we think really helps all Grain programmers, new or experienced, to look at Grain code and find it easy to read through because of the consistency, be than in end user code or the standard libraries.
```